doc/specs/#476 - Package Pinning.md

```diff
@@ -0,0 +1,205 @@
+---
+author: Yao Sun @yao-msft
+created on: 2022-10-12
+last updated: 2022-10-12
+issue id: 476
+---
+
+# Package Pinning
+
+For [#476](https://github.com/microsoft/winget-cli/issues/476)
+
+## Abstract
+
+This spec describes the functionality and high level implementation design of Package Pinning feature.
+
+## Inspiration
+
+This is inspired by functionalities in other package managers, as well as community feedback.
+- Packages may introduce breaking changes that users may not want integrate into their workflow quite yet.
+- Packages may update themselves so that it will be duplicate effort for winget to try to update them.
+- User may want to maintain some of the packages through other channels outside of winget, or prefer one source over others within winget.
+- User may want some of the packages to stay in some major versions but allow minor version changes during upgrade.
+
+## Solution Design
+
+#### Package Pinning types
+To achieve goals listed above, winget will support 3 types of Package Pinning:
+- **Blocking:** The package is blocked from `winget upgrade --all` or `winget upgrade <specific package>`, user has to unblock the package to let winget perform upgrade.
+- **Pinning:** The package is excluded from `winget upgrade --all` but allowed in `winget upgrade <specific package>`, a new argument `--include-pinned` will be introduced to let `winget upgrade --all` to include pinned packages.
+- **Gating:** The package is pinned to specific version(s). For example, if a package is pinned to version `1.2.*`, any version between `1.2.0` to `1.2.<anything>` is considered valid.
+
+To allow user override, `--force` can be used with `winget upgrade <specific package>` to override some of the pinning created above.
+
+#### Package Pinning Configuration Storage
+
+A separate sqlite db (other than the existing tracking catalog) will be created to store the package pinning configurations from user.
+```text
+PackageIdentifier       SourceIdentifier       Version        PinningType
+----------------------------------------------------------------------------
+Microsoft.TestApp       winget                 1.2.*          Gating
+```
+
+**Notes:** For this iteration, winget will only support pinning packages that are locally installed and correlatable with at least one of the remote sources. Winget will record a pinned package by the PackageIdentifier and SourceIdentifier. There can only be one pinning configuration for a specific package. In the future, winget may consider pinning packages from installed packages (upon improving the installed package's PackageIdentifier logic).
+
+## UI/UX Design
```

Comment:
The verbs `Add` and `Remove` are usually used for resources that support having multiple things added to them, such as with `Add-Content` appending to a file or `Add-ADGroupMember` adding one or more members to a group which may countain N amount of members total.

Since a package would probably only ever have one pin state (yes with options or no) at a time, I would maybe suggest the verbs `Suspend`/`Resume` (updates), `Get`/`Set`/`Reset` (a pin) or `Lock`/`Unlock` (a version),

Reply:
Ok the:
> There can only be one pinning configuration for a specific package.

threw me off here, I can see from further down the document now that a package can have multiple co-existing pins for different sources. That makes sense, so `Get`/`Add`/`Remove`/`Reset` (or Remove without parameters = Reset) makes sense

Reply:
Powershell support is not considered in this spec yet :(. I'll add it to future list and we'll probably get back to it after core feature is done.